package/MDAnalysis/analysis/pca.py

```diff
@@ -414,6 +414,130 @@ def transform(self, atomgroup, n_components=None, start=None, stop=None,
+
+        anchor : string, optional
+            The string to select the PCA atom whose displacement vector
+            is applied to non-PCA atoms in a residue. The atomselection
+            must have exactly one PCA atom in each residue of ``group``.
+            The default ``None`` does not extrapolate the projection
+            to non-PCA atoms.
+
+        Returns
+        -------
+        function
+            The resulting function f(ts) takes as input a
+            :class:`~MDAnalysis.coordinates.base.Timestep` ts,
+            and returns ts with the projected structure
+
+        Examples
+        --------
+        Run PCA class before using this function. For backbone PCA, run::
+
+            pca = PCA(universe, select='backbone').run()
+
+        Obtain a transformation function to project the
+        backbone trajectory onto the first principal component::
+
+            project = pca.project_single_frame(components=0)
+
+        To projet onto the first two components, run::
+
+            project = pca.project_single_frame(components=[0,1])
+
+        Alternatively, the transformation can be applied to PCA atoms and
+        extrapolated to other atoms according to the CA atom's translation
+        in each residue::
+
+            all = u.select_atoms('all')
+            project = pca.project_single_frame(components=0,
+                                               group=all, anchor='name CA')
+
+        Finally, apply the transformation function to a timestep::
+
+            project(u.trajectory.ts)
+
+        or apply the projection to the universe::
+
+            u.trajectory.add_transformations(project)
+
+
+        .. versionadded:: 2.2.0
+        """
+        if not self._calculated:
+            raise ValueError('Call run() on the PCA before projecting')
+
+        if group is not None:
+            if anchor is None:
+                raise ValueError("'anchor' cannot be 'None'" +
+                                 " if 'group' is not 'None'")
+
+            anchors = group.select_atoms(anchor)
+            anchors_res_ids = anchors.resindices
+            if np.unique(anchors_res_ids).size != anchors_res_ids.size:
+                raise ValueError("More than one 'anchor' found in residues")
+            if not np.isin(group.resindices, anchors_res_ids).all():
+                raise ValueError("Some residues in 'group'" +
+                                 " do not have an 'anchor'")
+            if not anchors.issubset(self._atoms):
+                raise ValueError("Some 'anchors' are not part of PCA class")
+
+            non_pca = group - self._atoms
+            pca_res_indices, pca_res_counts = np.unique(
+                self._atoms.resindices, return_counts=True)
+
+            # matrix_extrapolate is later multiplied to anchors' coordinates.
+            # multiplication shapes the array as appropriate for non_pca atoms
+            matrix_extrapolate = np.zeros((non_pca.n_atoms, anchors.n_atoms))
```

Comment:
Set the dtype to np.int8, the performance is comparable to the default allocation. int may improve memory efficiency though.
I tested a few types on my machine, and the default float64 is as fast as/faster than float32.

Reply:
You can even make the matrix Boolean, which makes the matrix multiplication a bit slower compared to int8, at least when I played with 100x100 matrices. 

For me, the fastest appears to be everything float32 or float64:
```
bool @ f64: 14.9 µs ± 278 ns
bool @ f32: 13.8 µs ± 735 ns
f64 @ f64 : 3.4 µs ± 143 ns
f32 @ f64:  11 µs ± 232 ns
f32 @ f32 : 3.56 µs ± 269 ns
i8 @ f32 : 11.1 µs ± 411 ns
i8 @ f64 : 13.3 µs ± 2.23 µs
```
Given that speed is an issue, I would say we do not care about memory but ensure that **there is no implicit type casting** involved: Ensure that **all terms in the matrix multiplication have the same type**.

Reply:
```python
import numpy as np
rng = np.random.default_rng()
M = rng.random((100, 100)) > 0.5
Mf = M.astype(np.float64)
Mf32 = M.astype(np.float32)
v = rng.random(100)
vf32 = v.astype(np.float32)

%timeit M @ v
%timeit Mf32 @ v32
```

Reply:
I see! I was playing with the dtype of only one matrix. No wonder, float32 was slower than float64, since the second matrix was implicitly defined as float64. Thank you!

Displacement vector seems to be in float32, so I've made `matrix_extrapolate` float32